File: api/elevenlabs_unlim.py
import requests
import time
import threading
from functools import wraps
from utils.settings import settings_manager
from utils.logger import logger, LogLevel

def retry(tries=3, delay=5, backoff=2):
    """
    A decorator for retrying a function or method if it fails.
    """
    def deco_retry(f):
        @wraps(f)
        def f_retry(*args, **kwargs):
            mtries, mdelay = tries, delay
            while mtries > 1:
                try:
                    return f(*args, **kwargs)
                except Exception as e:
                    msg = f"'{f.__name__}' failed with exception: {e}. Retrying in {mdelay} seconds..."
                    logger.log(msg, level=LogLevel.WARNING)
                    time.sleep(mdelay)
                    mtries -= 1
                    mdelay *= backoff
            return f(*args, **kwargs)
        return f_retry
    return deco_retry


# Use thread-local storage at module level to persist sessions across API instances
thread_local_storage = threading.local()

class ElevenLabsUnlimAPI:
    def __init__(self, api_key=None):
        self.api_key = api_key or settings_manager.get("elevenlabs_unlim_api_key")
        self.base_url = "https://elevenlabs-unlimited.net/api/v1"

    def _get_session(self):
        if not hasattr(thread_local_storage, "session"):
            thread_local_storage.session = requests.Session()
        return thread_local_storage.session

    def _make_request(self, method, endpoint, json=None, **kwargs):
        if not self.api_key:
            return None, "not_configured"
        
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }
        
        try:
            session = self._get_session()
            response = session.request(method, f"{self.base_url}/{endpoint}", headers=headers, json=json, timeout=30, **kwargs)
            
            # Special handling for 4xx errors to return status/message properly
            if response.status_code >= 400:
                try:
                    err_data = response.json()
                    detail = err_data.get('detail', response.text)
                    logger.log(f"API request to {endpoint} failed (Status {response.status_code}): {detail}", level=LogLevel.ERROR)
                    return err_data, "error"
                except:
                    logger.log(f"API request to {endpoint} failed with status {response.status_code}: {response.text}", level=LogLevel.ERROR)
                    return None, "error"
            
            if response.status_code == 200:
                if "audio/mpeg" in response.headers.get("Content-Type", ""):
                    return response.content, "connected"
                return response.json(), "connected"
            return response, "connected"

        except requests.exceptions.RequestException as e:
            logger.log(f"API request to {endpoint} failed: {e}", level=LogLevel.ERROR)
            return None, "error"

    def check_connection(self):
        logger.log("Checking ElevenLabsUnlim API connection...", level=LogLevel.INFO)
        _, status = self.get_balance()
        if status == "connected":
            logger.log("ElevenLabsUnlim API connection successful.", level=LogLevel.SUCCESS)
        else:
            logger.log("ElevenLabsUnlim API connection failed.", level=LogLevel.ERROR)
        return status

    @retry(tries=3, delay=5, backoff=2)
    def get_balance(self):
        # User stats endpoint: /api/v1/user/stats
        logger.log("Requesting ElevenLabsUnlim account stats...", level=LogLevel.INFO)
        data, status = self._make_request("get", "user/stats")
        if status == "connected" and data:
            # {
            #   "user_id": "user_123456",
            #   "subscription_type": "unlimited",
            #   "total_characters": 100000,
            #   "used_characters": 25000,
            #   "remaining_characters": 75000
            # }
            remaining = data.get("remaining_characters", 0)
            logger.log(f"Successfully retrieved balance: {remaining} chars", level=LogLevel.SUCCESS)
            return remaining, status
        
        if status == "not_configured":
             return None, status

        logger.log("Failed to retrieve ElevenLabsUnlim balance.", level=LogLevel.ERROR)
        return None, status
        
    def create_task(self, text, settings):
        """
        Create a voiceover task.
        settings dict should contain:
        - voice_id
        - model_id (optional, default 'eleven_multilingual_v2')
        - stability, similarity_boost, style, use_speaker_boost inside 'voice_settings'
        """
        logger.log("Creating ElevenLabsUnlim voiceover task...", level=LogLevel.INFO)
        
        payload = {
            "text": text,
            "voice_id": settings.get("voice_id", "21m00Tcm4TlvDq8ikWAM"), # default to Rachel
            "model_id": settings.get("model_id", "eleven_multilingual_v2"),
            "voice_settings": {
                "stability": settings.get("stability", 0.5),
                "similarity_boost": settings.get("similarity_boost", 0.75),
                "style": settings.get("style", 0.0),
                "use_speaker_boost": settings.get("use_speaker_boost", True)
            }
        }
        
        data, status = self._make_request("post", "voice/synthesize", json=payload)
        # Expected: {"task_id": "...", "status": "pending", ...}
        
        if status == "connected" and data and "task_id" in data:
            task_id = data.get("task_id")
            logger.log(f"Successfully created task with ID: {task_id}", level=LogLevel.SUCCESS)
            return task_id, status
            
        logger.log("Failed to create ElevenLabsUnlim task.", level=LogLevel.ERROR)
        return None, "error"

    def get_task_status(self, task_id):
        # GET /api/v1/voice/status/{task_id}
        # Returns 200 OK with status string or json? Doc says response value schema "string". Assume it returns JSON with status inside?
        # Actually doc example for status endpoint response is "string", but typically it's json.
        # Let's inspect create_task response: "status": "pending".
        # Let's assume it returns a JSON object with status. If not, we handle string.
        # Wait, doc says Response Schema: "string". Let's assume the body IS the status string.
        
        data, status = self._make_request("get", f"voice/status/{task_id}")
        
        # If the API returns a raw string for status, _make_request might parse it as json if it's quoted?
        # Or if it's just a string, requests.json() might fail if not valid json.
        # However, _make_request returns response.json() if 200.
        
        if status == "connected":
            # If data is a dict, look for status. If it's a string, it IS the status.
            task_status = data.get("status") if isinstance(data, dict) else data
            return task_status, status
            
        return None, status

    def get_task_result(self, task_id):
        # GET /api/v1/voice/download/{task_id}
        logger.log(f"Downloading audio for task {task_id}...", level=LogLevel.INFO)
        
        # We need the raw content here
        if not self.api_key:
            return None, "not_configured"
            
        headers = {"Authorization": f"Bearer {self.api_key}"}
        url = f"{self.base_url}/voice/download/{task_id}"
        
        try:
            session = self._get_session()
            response = session.get(url, headers=headers, timeout=60)
            
            if response.status_code == 200:
                logger.log(f"Successfully downloaded audio for task {task_id}", level=LogLevel.SUCCESS)
                return response.content, "connected"
            else:
                 logger.log(f"Failed to download audio for task {task_id}. Status: {response.status_code}", level=LogLevel.ERROR)
                 return None, "error"
                 
        except requests.exceptions.RequestException as e:
            logger.log(f"Request failed for task {task_id} download: {e}", level=LogLevel.ERROR)
            return None, "error"

File: backend/api/elevenlabsunlim.go
package api

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"soloveyko/backend/utils"
	"time"
)

type ElevenLabsUnlimService struct {
	settings *utils.SettingsService
	baseUrl  string
	OnLog    func(level string, message string, details ...string)
}

func NewElevenLabsUnlimService(settings *utils.SettingsService) *ElevenLabsUnlimService {
	return &ElevenLabsUnlimService{
		settings: settings,
		baseUrl:  "https://voicer.mat3u.com/api/v1",
	}
}

type VoicerStatsResponse struct {
	SubscriptionType    string `json:"subscription_type"`
	TotalCharacters     int    `json:"total_characters"`
	UsedCharacters      int    `json:"used_characters"`
	RemainingCharacters int    `json:"remaining_characters"`
}

// GetBalance отримує поточний баланс користувача
func (s *ElevenLabsUnlimService) GetBalance(apiKey string) (float64, error) {
	if apiKey == "" {
		return 0, fmt.Errorf("API key is empty")
	}

	client := &http.Client{Timeout: 30 * time.Second}
	req, err := http.NewRequest("GET", s.baseUrl+"/user/stats", nil)
	if err != nil {
		return 0, err
	}

	req.Header.Set("Authorization", "Bearer "+apiKey)

	resp, err := client.Do(req)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return 0, fmt.Errorf("API error: %d", resp.StatusCode)
	}

	var stats VoicerStatsResponse
	if err := json.NewDecoder(resp.Body).Decode(&stats); err != nil {
		return 0, err
	}

	if stats.SubscriptionType == "unlimited" {
		return -1, nil
	}

	return float64(stats.RemainingCharacters), nil
}

type VoicerSynthesizeRequest struct {
	Text               string                 `json:"text"`
	VoiceID            string                 `json:"voice_id,omitempty"`
	ModelID            string                 `json:"model_id,omitempty"`
	SplitType          string                 `json:"split_type,omitempty"`
	SplitOutput        bool                   `json:"split_output,omitempty"`
	AutoPauseEnabled   bool                   `json:"auto_pause_enabled,omitempty"`
	AutoPauseDuration  float64                `json:"auto_pause_duration,omitempty"`
	AutoPauseFrequency int                    `json:"auto_pause_frequency,omitempty"`
	VoiceSettings      map[string]interface{} `json:"voice_settings,omitempty"`
}

type VoicerSynthesizeResponse struct {
	TaskID  string `json:"task_id"`
	Status  string `json:"status,omitempty"`
	Message string `json:"message,omitempty"`
}

type VoicerStatusResponse struct {
	Status    string  `json:"status"`
	Progress  float64 `json:"progress,omitempty"`
	ErrorText string  `json:"error,omitempty"`
}

func (s *ElevenLabsUnlimService) CreateTask(apiKey string, text string, voiceID string, voiceSettings map[string]interface{}) (string, error) {
	reqBody := VoicerSynthesizeRequest{
		Text:          text,
		VoiceID:       voiceID,
		ModelID:       "eleven_multilingual_v2",
		VoiceSettings: voiceSettings,
	}

	jsonData, err := json.Marshal(reqBody)
	if err != nil {
		return "", err
	}

	// Збільшено таймаут до 120 секунд для створення задачі
	client := &http.Client{Timeout: 120 * time.Second}
	req, err := http.NewRequest("POST", s.baseUrl+"/voice/synthesize", bytes.NewBuffer(jsonData))
	if err != nil {
		return "", err
	}

	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusAccepted && resp.StatusCode != http.StatusCreated {
		return "", fmt.Errorf("API error %d: %s", resp.StatusCode, string(body))
	}

	var res VoicerSynthesizeResponse
	if err := json.Unmarshal(body, &res); err != nil {
		return "", fmt.Errorf("failed to parse response: %v | Body: %s", err, string(body))
	}

	return res.TaskID, nil
}

func (s *ElevenLabsUnlimService) GetTaskStatus(apiKey string, taskID string) (string, error) {
	client := &http.Client{Timeout: 30 * time.Second}
	req, err := http.NewRequest("GET", fmt.Sprintf("%s/voice/status/%s", s.baseUrl, taskID), nil)
	if err != nil {
		return "", err
	}

	req.Header.Set("Authorization", "Bearer "+apiKey)

	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("API error %d: %s", resp.StatusCode, string(body))
	}

	var res VoicerStatusResponse
	if err := json.Unmarshal(body, &res); err == nil && res.Status != "" {
		if res.Status == "failed" {
			if res.ErrorText != "" {
				return "failed", fmt.Errorf("API error: %s", res.ErrorText)
			}
			return "failed", nil
		}
		return res.Status, nil
	}

	var statusStr string
	if err := json.Unmarshal(body, &statusStr); err == nil {
		return statusStr, nil
	}

	return string(bytes.Trim(body, "\" ")), nil
}

func (s *ElevenLabsUnlimService) DownloadResult(apiKey string, taskID string, filePath string) error {
	client := &http.Client{Timeout: 600 * time.Second}
	req, err := http.NewRequest("GET", fmt.Sprintf("%s/voice/download/%s", s.baseUrl, taskID), nil)
	if err != nil {
		return err
	}

	req.Header.Set("Authorization", "Bearer "+apiKey)

	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("API error %d: %s", resp.StatusCode, string(body))
	}

	out, err := os.Create(filePath)
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, resp.Body)
	return err
}

func (s *ElevenLabsUnlimService) Synthesize(apiKey string, text string, voiceID string, voiceSettings map[string]interface{}, outputPath string, id string, taskLabel string) error {
	if s.OnLog != nil {
		s.OnLog("INFO", "[ElevenLabsUnlim] Starting voice synthesis...", id, taskLabel)
	}

	taskID, err := s.CreateTask(apiKey, text, voiceID, voiceSettings)
	if err != nil {
		return err
	}

	if s.OnLog != nil {
		s.OnLog("INFO", fmt.Sprintf("[ElevenLabsUnlim] Task created: %s. Polling status...", taskID), id, taskLabel)
	}

	maxAttempts := 120
	for i := 0; i < maxAttempts; i++ {
		status, err := s.GetTaskStatus(apiKey, taskID)
		if err != nil {
			return err
		}

		if s.OnLog != nil {
			s.OnLog("INFO", fmt.Sprintf("[ElevenLabsUnlim] Task %s status: %s", taskID, status), id, taskLabel)
		}

		switch status {
		case "completed":
			if s.OnLog != nil {
				s.OnLog("INFO", "[ElevenLabsUnlim] Synthesis completed. Downloading...", id, taskLabel)
			}
			return s.DownloadResult(apiKey, taskID, outputPath)
		case "failed":
			return fmt.Errorf("synthesis failed (status: failed)")
		}

		time.Sleep(5 * time.Second)
	}

	return fmt.Errorf("synthesis timeout")
}

func (s *ElevenLabsUnlimService) SaveAPIKey(apiKey string) error {
	return s.settings.SetElevenLabsUnlimAPIKey(apiKey)
}

func (s *ElevenLabsUnlimService) GetAPIKey() string {
	return s.settings.GetElevenLabsUnlimAPIKey()
}
